### ISSUE TITLE
Contracts type Inference / autocompletions (& custom hooks vs wagmi hooks)

### ISSUE THREAD

**Issue by technophile-04** [scaffold-eth/se-2#116]
Currently, we are trying to create some utilities/hooks  as Carlos mentioned in #88 : 
> When creating your app in SE, it's super handy to have hooks that loads the deployed contracts (hardhat_contracts / external_contracts).
In SE-2 we are using wagmi, but I think we could benefit of having some layer (hooks / utils) on top of wagmi to load these deployed contracts.

A basic iteration of `useScaffoldContractRead` and `useScaffoldContractWrite` has been created at #93 but it's very loosely typed....The reason for creating this utility is if you use wagmi's `useContractRead/Write` you always need to pass in `abi` and `address` but we can handle it internally. 

What we want to achieve is something like this -> https://twitter.com/wagmi_sh/status/1580591788611338240 with the above hooks. 

We can leverage typescript generics as mentioned in this comment https://github.com/scaffold-eth/se-2/pull/93#issuecomment-1383935227 and even [`useContractRead`](https://github.com/wagmi-dev/wagmi/blob/6f7dacfe5e09e01eb8c2c73ed33db9ce2e2f1223/packages/react/src/hooks/contracts/useContractRead.ts#L75) accepts generics. 

Some resources / usefull lookups : 
1) Problem faced while implementing it : https://github.com/scaffold-eth/se-2/pull/93#issue-1534650952
2) https://github.com/scaffold-eth/se-2/pull/93#issuecomment-1403947795
3) https://github.com/scaffold-eth/se-2/pull/93#issuecomment-1403996177

**Comment by carletex** [comment]
Not sure if useful, but there is a `typechain-types` folder generated inside the `hardhat` package when you compile/deploy a contract.

**Comment by technophile-04** [comment]
I have been exploring different ways to conquer this and here is my research : 
- The way wagmi handles the autocompletes is using their `ABIType` packagage and they recommend to use `abi as const` or inline the abi checkout -> https://wagmi.sh/react/typescript#type-inference

I think the approach taken by @rin-st in #147  seems correct to me, although the developer needs to pass in the `abi` manually to `useScaffoldContractRead`, I feel this is the only way to get autocompletion for `functionName` as shown by Rinat in https://github.com/scaffold-eth/se-2/pull/147#issuecomment-1424839258

I also tried exploring `typechain-types` and I think it works great with `useContract` hook but then we need to handle all the states like `loading, isError` etc which `useContractRead/write` provides. checkout -> https://github.com/wagmi-dev/wagmi/discussions/853

### using [@wagmi-cli](https://twitter.com/wagmi_sh/status/1620097176738357248)
I think was just launched 10 days before seems like a great solution for write and read hooks. checkout example video  https://www.youtube.com/watch?v=sBX79sPjRss&t=219s. 

I have also created a branch [`feat/wagmi-cli`](https://github.com/scaffold-eth/se-2/tree/feat/wagmi-cli) where I am using hooks generated by cli in Example UI and found them really good. 

To use [`feat/wagmi-cli`](https://github.com/scaffold-eth/se-2/tree/feat/wagmi-cli) branch : 
1. checkout to branch  [`feat/wagmi-cli`](https://github.com/scaffold-eth/se-2/tree/feat/wagmi-cli)
2. run `yarn install`
3. `yarn chain`
4. `yarn deploy`
5. `yarn wagmi generate` to generate hooks inside `packages/nextjs/generated`
6. `yarn start` and checkout http://localhost:3000/example-ui

would love to hear thoughts on @wagmi-cli 

cc @carletex  @rin-st

**Comment by rin-st** [comment]
Excellent research, @technophile-04 ! `wagmi-cli` is awesome!

Since it adds autocomplete and works great with other wagmi types, I believe it's all we need. This solution adds a new step `yarn wagmi generate`, but we can add it let's say to `deploy` script.

My solution needs to add lots of types to work well with configs, return types etc and I didn't find fast ways to complete it.
Typechain and solution from se-ts doesn't sweet us too, since
- first we need to add types manually everywhere and as I understand and as you mentioned it does not work well with wagmi
- second adds old lib https://github.com/dethcrypto/eth-sdk and I believe we don't need it

**Comment by sverps** [comment]
Not sure what I'm doing wrong, but when I try the `yarn wagmi generate` step, it fails with following error:
```
yarn wagmi generate
⠸ Validating plugins
hardhat must be installed to use Hardhat plugin.
To install, run: npm install --save hardhat
```

**Comment by rin-st** [comment]
It doesn't work for other people too, but for some reason works for me. Probably you need to move `hardhat` from `devDependencies` to `dependencies` in `packages/hardhat/package.json` or recreate `yarn.lock` :shrug:

**Comment by sverps** [comment]
Tried installing hardhat globally, tried removing yarn.lock and tried different places to put the dependency, but I keep getting the same issue.

**Comment by technophile-04** [comment]
Tysm guys, and yup sorry don't why it worked for me initially :( ...I am also getting the same error on `hardhat must be installed to use Hardhat plugin.` on the fresh clone. 

I have raised an issue to Wagmi about this. I think we can keep `wagmi-cli` as `good to have` and come back to it late for now since we are almost there with native auto-completes thanks to https://github.com/scaffold-eth/se-2/pull/147#issuecomment-1435779143

PS: Although if someone wants to tinker...you can do `npm install --save hardhat` instead of `yarn` in `packages/hardhat` and run `yarn install` after it. This is hack which worked for me but not at all seems like an feasible solution

**Comment by carletex** [comment]
Since this is going to take some time to get it right, I changed it from "beta launch" (1 week away :D) to "enhancement".

What we definitely need is #189

**Comment by technophile-04** [comment]
PS : The issue was resolved and updated [feat/wagmi-cli](https://github.com/scaffold-eth/se-2/tree/feat/wagmi-cli), if someone wants to try the instruction remains same -> https://github.com/scaffold-eth/se-2/issues/116#issuecomment-1426819707

Don't want divert us from #189 we are almost there and as Carlos mentioned its really new and lets keep it as `enhancement`

**Comment by carletex** [comment]
> Don't want divert us from https://github.com/scaffold-eth/se-2/pull/147 we are almost there and as Carlos mentioned its really new and lets keep it as enhancement

We should definitely explore it on a parallel PR! Feel free to open a draft PR with this @technophile-04 , so we can test and see what look best.... and decide.

We are not in a rush now that #189 was solved :)

**Comment by carletex** [comment]
We are working on:
- https://github.com/scaffold-eth/se-2/pull/200
- https://github.com/scaffold-eth/se-2/pull/202

Both are looking great. 

**Open questions**: 

1. Are they exclusive? Like, if we use wagmi hooks to interact with contracts, should we remove our custom hooks (scaffoldContractRead / write). The question also work the other way around :)

2. Should we wait after Denver before merging? I want `main` to be super stable on Denver. Maybe we can tell people to use the branches to test.

cc @technophile-04 @rin-st @sverps

**Comment by sverps** [comment]
1. Not exclusive per se, but I'm not sure if it improves DX. It might just cause confusion that there's two distinct ways to interact with contracts.
2. I think it's fine to wait, especially since there's little time to play a bit with the two options and figure out the benefits and downsides of each.

## Some thoughts already

On the custom types approach:
- It makes our custom hooks a bit more complex. Remaining to test that it properly works in all scenario's.
- It allow for very flexible usage: users just need to remember 3 hooks to read/write or access events, then fill in the contract details via autocomplete.

On the wagmi-cli approach:
- It's clean, but has the downside that it's a bit less intuitive to build generic contract components. But I'm not sure if we really need generic components except for the debug page, which we can manage.
- It has the added advantage of externalizing these contract hooks, which will make it easier to maintain se-2 (since potential bugs in their generator will get fixed at wagmi side)
- Downside to this last point is that we might need to wait for issues to get resolved

**Comment by carletex** [comment]
Great insights @sverps!! I agree 100 % with everything you said. Let's keep discussing and playing with both options.

P.S. We are going to talk with the wagmi people at some point in March. So we can prepare questions / requests :D 

----

One more thing to consider: **external contracts**. Sometimes you want to interact with external contracts (e.g. get the DAI balance of a given address). Options:

1. Use **wagmi** directly with the contract address / abi: useContractRead / useContractWrite
2. Use our **custom hooks**: Right now only possible on `useScaffoldContractRead` where you can override stuff (address, abi, chainId) with `readConfig`.
3. Use ethers directly! It might seem that it doesn't make sense, but we had to use it yesterday on the hack session. Maybe there was an alternative that we missed.

1 seems more appropiate? (when you don't have to use 3  :D)

## Why we had to use ethers directly
(Happy to create a new issue for this)

We had a list of addresses and we wanted to get the ERC20 balance for each address (calling the smart contract `getBalance` for each). With hooks (& its declarative-way) we couldn't do that, so we had to [instantiate the contract with ethers](https://docs.ethers.org/v5/api/contract/contract/#Contract--creating) so we could loop over the addresses (in a useEffect when `addressed` changed), call contract.getBalance(address) and store the balance for each address.

Did we miss something?

**Comment by technophile-04** [comment]
> Great insights @sverps!! I agree 100 % with everything you said.

++ 💯

> One more thing to consider: external contracts. Sometimes you want to interact with external contracts (e.g. get the DAI balance of a given address). Options:

wagmi-cli does seem to have `plugin` to  generate hooks straight from block explorer checkout -> https://wagmi.sh/cli/plugins/blockExplorer

**Comment by rin-st** [comment]
> With hooks (& its declarative-way) we couldn't do that

Have you tried https://wagmi.sh/react/hooks/useContractReads ? We can create a new wrapper for it or extend `useScaffoldContractRead`

---

> Alternatives are:

https://github.com/scaffold-eth/se-2/pull/147#issuecomment-1437373694
> The best solution would be that the generation happens on the build step, i.e. running the hardhat node + deploy on build.
----

> On the wagmi-cli approach
> - It has the added advantage of externalizing these contract hooks, which will make it easier to maintain se-2 (since potential bugs in their generator will get fixed at wagmi side)

As I understand it's a big advantage. Yes, it creates some chaos with all that tons of hooks but I believe with proper docs it will be much better for dx and for the end users too.

But again, for
> With hooks (& its declarative-way) we couldn't do that

there is no solution with generated hooks 🤷 .

So again I don't know which approach is better 😄 , probably custom types

**Comment by technophile-04** [comment]
Thanks @sverps, 

> Alternatives are:

I like option 1, it seems minimal and good. 

Even option 2 seems great but for running `yarn deploy` you need `yarn chain` running.
If we do `yarn chain && yarn deploy` this won't run `yarn deploy`. 

#### Some alternatives to get it to work : 

1. To achieve we may need to use something like [ttab](https://www.npmjs.com/package/ttab) which helps you run commands in new terminal window, but **ttab** does not work with windows.  

2. Running them concurrently using `yarn chain & yarn deploy` notice there is only one `&` I tried this option but dosen't work with my terminal might be an issue from my side. 

also instead of `pre:start` we can use `prestart` as npm natively supports it checkout -> [npm docs](https://docs.npmjs.com/cli/v9/using-npm/scripts#pre--post-scripts)

As I didn't find a sure-shot way to run both `yarn chain` and `yarn deploy`, I am a bit biased towards option 1 plus it seems easy and minimal, Pardon me If I missed something. 

--- 

After some thinking my views on `wagmi-cli` and `custom-hooks` co-existence : 

> Yes, it creates some chaos with all that tons of hooks but I believe with proper docs it will be much better for dx and for the end users too.

💯, adding to this even though there are some intersection b/w them, there also have some DX advantages 

eg : If you want to itneract with external contract then instead of going to etherscan, copying abi, putting abi locally is tassk, instead you can easily generate hooks by just providing contract address and etherscan api key using wagmi-cli [plugins](https://wagmi.sh/cli/plugins/etherscan)

#### Also some people might be overwhelmed with hooks created from `wagmi-cli` if they haven't touched the contract and don't remember all the function names. with `custom-hooks` you get quick and short lookup of all functions names you can call. 

Also if we use both together then we don't need to write our own script to generate abi's as `.ts` file and `as const`. `wagmi-cli` does this for us which we can leverage. Also it creates helper configs which we can leverage for types autocompleteions for our custom hooks 

```js
export const yourContractAddress = {
  5: "0x60fD3D8C0a7d6c4E4b54B8591333952C8cbC3AC1",
} as const;

export const yourContractConfig = { address: yourContractAddress, abi: yourContractABI } as const;
```

**Comment by sverps** [comment]
Good points. After playing a bit with it, I'm leaning slightly towards the custom hooks. The point about having wagmi-cli generate hooks for external conracts is cool though.

> also instead of `pre:start` we can use `prestart` as npm natively supports it checkout -> [npm docs](https://docs.npmjs.com/cli/v9/using-npm/scripts#pre--post-scripts)

Since se-2 is `yarn` based, this won't work. Yarn 2 has deprecated this functionality ([yarn docs](https://yarnpkg.com/advanced/lifecycle-scripts)), so if we want to add something like that, we'll need to call it manually.

**Comment by technophile-04** [comment]
>  After playing a bit with it, I'm leaning slightly towards the custom hooks. The point about having wagmi-cli generate hooks for external conracts is cool though.
 
 Yup, I think using write functions generated from `wagmi-cli` is PITA. 
 
![transactor-cli](https://user-images.githubusercontent.com/80153681/221370090-7f52c6d6-088d-40ce-a289-b11a3aae3932.png)


That said, I think we can suggest to  use our `custom hooks` in docs and also use them in `Example UI` and maybe we can have a small side section where we say we have `wagmi-cli` inbuilt and we can refer to some plugins and  for external contract stuff. Also wagmi-cli  will help us get generated files for `custom-hooks` 

> Since se-2 is yarn based, this won't work. Yarn 2 has deprecated this functionality ([yarn docs](https://yarnpkg.com/advanced/lifecycle-scripts)), so if we want to add something like that, we'll need to call it manually.

Yup my bad, sorry didn't see it 🙌

**Comment by carletex** [comment]
bumping this:

It seems that we are going with custom hooks (auto completion + types) from https://github.com/scaffold-eth/se-2/pull/202, and we are waiting on https://github.com/scaffold-eth/se-2/pull/231 so we can get `export type DefaultChain = "31337";` dynamically, right?

cc @technophile-04 @sverps

**Comment by sverps** [comment]
Sounds good :+1: 
I'll update #202 and prepare it for the changes in #231 and #191. As soon as those are completed, I'll then be able to finalize this one.
